### PR TITLE
MULE-14131: No parser found for compatibility component if declared in domain

### PR DESCRIPTION
**Pull request by lucianoRM**

### CODE REVIEW

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/AbstractDeploymentTestCase.java [hunk 1]
```diff
@@ -1180,6 +1185,47 @@ private ArtifactPluginFileBuilder createHelloExtensionV2PluginFileBuilder() {
         .describedBy((mulePluginModelBuilder.build()));
   }
 
+  private ArtifactPluginFileBuilder createExceptionThrowingPluginFileBuilder() {
+    MulePluginModel.MulePluginModelBuilder mulePluginModelBuilder = new MulePluginModel.MulePluginModelBuilder()
+        .setMinMuleVersion(MIN_MULE_VERSION)
+        .setName("exceptionPlugin")
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/AbstractDeploymentTestCase.java]: Extract to a local constant

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/ApplicationPolicyDeploymentTestCase.java [hunk 2]
```diff
@@ -296,7 +300,45 @@ public void appliesApplicationPolicyDuplicatingPlugin() throws Exception {
                                                       getResourceFile("/appPluginPolicy.xml"), emptyList()));
 
     executeApplicationFlow("main");
-    assertThat(invocationCount, equalTo(1));
+  }
+
+  @Test
+  public void appliesApplicationPolicyDuplicatingPluginWithNoSDKHandledPlugin() throws Exception {
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/ApplicationPolicyDeploymentTestCase.java]: NoSDKHandledPlugin -> Plugin

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/ApplicationPolicyDeploymentTestCase.java [hunk 3]
```diff
@@ -296,7 +300,45 @@ public void appliesApplicationPolicyDuplicatingPlugin() throws Exception {
...
+    startDeployment();
+    assertApplicationDeploymentSuccess(applicationDeploymentListener, applicationFileBuilder.getId());
+
+    policyManager.addPolicy(applicationFileBuilder.getId(), exceptionPolicyFileBuilder.getArtifactId(),
+                            new PolicyParametrization(policyName, s -> true, 1, emptyMap(),
+                                                      getResourceFile("/exceptionThrowingPolicy.xml"), emptyList()));
+    try {
+      executeApplicationFlow("main");
+      fail();
+    } catch (MuleRuntimeException e) {
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/ApplicationPolicyDeploymentTestCase.java]: e -> expected

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DomainDeploymentTestCase.java [hunk 4]
```diff
@@ -83,6 +84,9 @@
       new DomainFileBuilder("shared-domain").definedBy("shared-domain-config.xml");
   private final DomainFileBuilder domainWithPropsFileBuilder =
       new DomainFileBuilder("domain-with-props").definedBy("domain-with-props-config.xml");
+  private final DomainFileBuilder exceptionThrowingPluginImportingDomain =
+      new DomainFileBuilder("exception-throwi-plugin-importing-domain").definedBy("empty-domain-config.xml")
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DomainDeploymentTestCase.java]: exception-throwi-plugin-importing-domain > exception-throwing-plugin-importing-domain

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DomainDeploymentTestCase.java [hunk 5]
```diff
@@ -204,6 +208,24 @@ public void deploysDomainWithSharedLibPrecedenceOverApplicationPluginLib() throw
     executeApplicationFlow("main");
   }
 
+  @Test
+  public void pluginNotHandledBySDKFromDomainUsedInApp() throws Exception {
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DomainDeploymentTestCase.java]: Rename test

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/ApplicationPolicyDeploymentTestCase.java [hunk 6]
```diff
@@ -296,7 +300,45 @@ public void appliesApplicationPolicyDuplicatingPlugin() throws Exception {
...
+
+    startDeployment();
+    assertApplicationDeploymentSuccess(applicationDeploymentListener, applicationFileBuilder.getId());
+
+    policyManager.addPolicy(applicationFileBuilder.getId(), exceptionPolicyFileBuilder.getArtifactId(),
+                            new PolicyParametrization(policyName, s -> true, 1, emptyMap(),
+                                                      getResourceFile("/exceptionThrowingPolicy.xml"), emptyList()));
+    try {
+      executeApplicationFlow("main");
+      fail();
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/ApplicationPolicyDeploymentTestCase.java]: Add cause of failure

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DomainDeploymentTestCase.java [hunk 7]
```diff
@@ -204,6 +208,24 @@ public void deploysDomainWithSharedLibPrecedenceOverApplicationPluginLib() throw
...
+    addPackedDomainFromBuilder(exceptionThrowingPluginImportingDomain);
+
+    ApplicationFileBuilder applicationFileBuilder =
+        createExtensionApplicationWithServices("exception-throwing-app.xml").dependingOn(exceptionThrowingPluginImportingDomain);
+    addPackedAppFromBuilder(applicationFileBuilder);
+    startDeployment();
+
+    try {
+      executeApplicationFlow("main");
+      fail();
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DomainDeploymentTestCase.java]: Add cause of failure

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DomainDeploymentTestCase.java [hunk 8]
```diff
@@ -204,6 +208,24 @@ public void deploysDomainWithSharedLibPrecedenceOverApplicationPluginLib() throw
...
+    } catch (MuleRuntimeException e) {
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DomainDeploymentTestCase.java]: e -> expected

File: modules/deployment/src/test/resources/exception/META-INF/mule-exception.xsd [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+    <xsd:import namespace="http://www.w3.org/XML/1998/namespace"/>
+    <xsd:import namespace="http://www.mulesoft.org/schema/mule/core"
+                schemaLocation="http://www.mulesoft.org/schema/mule/core/current/mule.xsd"/>
+
+
+    <xsd:element name="throw-exception" type="throwType" substitutionGroup="mule:abstract-message-processor">
+        <xsd:annotation>
+            <xsd:documentation>
+                Mock message source that links to a shared-config.
```
Comment (pablokraan) [inline modules/deployment/src/test/resources/exception/META-INF/mule-exception.xsd]: Is this description correct? (looks like a copy/paste from another file)

File: modules/deployment/src/test/resources/exception/META-INF/mule-exception.xsd [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+            </xsd:documentation>
+        </xsd:annotation>
+    </xsd:element>
+
+    <xsd:complexType name="throwType">
+        <xsd:complexContent>
+            <xsd:extension base="mule:abstractMessageProcessorType">
+                <xsd:attribute name="exception" type="mule:substitutableName" use="required">
```
Comment (pablokraan) [inline modules/deployment/src/test/resources/exception/META-INF/mule-exception.xsd]: exceptionClassName or className?

File: modules/deployment/src/test/resources/org/exception/ExceptionComponentBuildingDefinitionProvider.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+public class ExceptionComponentBuildingDefinitionProvider implements ComponentBuildingDefinitionProvider {
+
+  private ComponentBuildingDefinition.Builder baseDefinition;
+
+  public ExceptionComponentBuildingDefinitionProvider() {}
```
Comment (pablokraan) [inline modules/deployment/src/test/resources/org/exception/ExceptionComponentBuildingDefinitionProvider.java]: Is this constructor needed?

File: modules/deployment/src/test/resources/org/exception/ExceptionComponentBuildingDefinitionProvider.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+    componentBuildingDefinitions.add(baseDefinition
+                                             .withIdentifier("throw-exception")
+                                             .withTypeDefinition(fromType(ThrowProcessor.class))
+                                             .withSetterParameterDefinition("exception", fromSimpleParameter("exception").build())
+                                             .withSetterParameterDefinition("error", fromSimpleParameter("error").build())
+                                             .withSetterParameterDefinition("count", fromSimpleParameter("count").withDefaultValue(-1).build()).build());
+
+    return componentBuildingDefinitions;
+
```
Comment (pablokraan) [inline modules/deployment/src/test/resources/org/exception/ExceptionComponentBuildingDefinitionProvider.java]: Remove extra lines

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/SpringXmlConfigurationBuilder.java [hunk 13]
```diff
@@ -150,15 +149,30 @@ private MuleArtifactContext doCreateApplicationContext(MuleContext muleContext,
...
 
     return new MuleArtifactContext(muleContext, resolveArtifactConfigResources(), artifactDeclaration, optionalObjectsController,
-                                   getArtifactProperties(), artifactType, resolveContextArtifactPluginClassLoaders(),
+                                   getArtifactProperties(), artifactType, resolveAppAndDomainPluginClassLoaders(),
                                    resolveParentConfigurationProperties(), disableXmlValidations);
   }
 
+  private List<ClassLoader> resolveAppAndDomainPluginClassLoaders() {
+    List<ClassLoader> artifactContextPluginClassLoaders = new ArrayList<>(resolveContextArtifactPluginClassLoaders());
+    if (artifactType.equals(ArtifactType.APP) && parentContext != null) {
```
Comment (pablokraan) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/SpringXmlConfigurationBuilder.java]: If I remember correctly, this check is related to a previous bug you found. Add the jira here

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/SpringXmlConfigurationBuilder.java [hunk 14]
```diff
@@ -150,15 +149,30 @@ private MuleArtifactContext doCreateApplicationContext(MuleContext muleContext,
...
                                          resolveParentConfigurationProperties(), disableXmlValidations);
     }
 
     return new MuleArtifactContext(muleContext, resolveArtifactConfigResources(), artifactDeclaration, optionalObjectsController,
-                                   getArtifactProperties(), artifactType, resolveContextArtifactPluginClassLoaders(),
+                                   getArtifactProperties(), artifactType, resolveAppAndDomainPluginClassLoaders(),
                                    resolveParentConfigurationProperties(), disableXmlValidations);
   }
 
+  private List<ClassLoader> resolveAppAndDomainPluginClassLoaders() {
```
Comment (pablokraan) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/SpringXmlConfigurationBuilder.java]: I'm failing to see how this method return the Domain plugin classloaders. Can you explain how that works?

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/SpringXmlConfigurationBuilder.java (hunk 13 above)
Comment (lucianoRM) [reply modules/spring-config/src/main/java/org/mule/runtime/config/internal/SpringXmlConfigurationBuilder.java]: Actually, the logic is wrong. We should do this when the parentContext is not null.

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/SpringXmlConfigurationBuilder.java (hunk 14 above)
Comment (lucianoRM) [reply modules/spring-config/src/main/java/org/mule/runtime/config/internal/SpringXmlConfigurationBuilder.java]: I renamed the method to resolveArtifactAndParentContextsPluginClassLoaders() since it's also executed for policies and apps.

For what i could see, the imported static method (resolveContextArtifactPluginClassLoaders()) calls "getArtifactPluginClassLoaders" on the current thread contextClassLoader. Getting the pluginClassLoaders defined when that particular contextClassLoader was created. (Check org.mule.runtime.module.deployment.impl.internal.application.DefaultApplicationFactory#createArtifact).

So by getting the parentContext's classLoader and setting it as the current thread context one, i end up getting the domain class loader (when the artifact is a policy, i get the app class loader), and the static method looks for the plugins class loaders from de domain one.

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/SpringXmlConfigurationBuilder.java (hunk 13 above)
Comment (lucianoRM) [reply modules/spring-config/src/main/java/org/mule/runtime/config/internal/SpringXmlConfigurationBuilder.java]: Actually, the logic is wrong. I should do this for when parentContext is not null

File: modules/artifact/src/test/java/org/mule/runtime/module/artifact/builder/AbstractArtifactFileBuilder.java [hunk 15]
```diff
@@ -193,7 +193,11 @@ public File getArtifactFile() {
         final List<ZipResource> zipResources = new LinkedList<>(resources);
         zipResources.add(new ZipResource(getArtifactPomFile().getAbsolutePath(), getArtifactFileBundledPomPath()));
         zipResources.addAll(getCustomResources());
-        compress(tempFile, zipResources.toArray(new ZipResource[0]));
+        try {
+          compress(tempFile, zipResources.toArray(new ZipResource[0]));
+        } catch (Exception e) {
+          e.printStackTrace();
```
Comment (pablokraan) [inline modules/artifact/src/test/java/org/mule/runtime/module/artifact/builder/AbstractArtifactFileBuilder.java]: Why this change?
